### ISSUE TITLE
Write a cool pretty printer for an `Encoding`

### ISSUE THREAD

**Issue by thoughtpolice** [well-typed/binary-serialise-cbor#35]
One of the nicest things about this library is that the `Encoding` type are really just "deep embeddings", or as we like to call them: syntax trees. This allows a variety of 'fun' things.

When you get an `Encoding`, that's really something like a function `Tokens -> Tokens`, and you apply a `TkEnd` to get a `Tokens` you can traverse over recursively.

It would be neat if we had a way to 'pretty print' this `Encoding` into a representation of what will actually be a CBOR value. This would be quite useful for visualising how a particular data type would be encoded into CBOR, or merely how it's structurally represented internally.

**Comment by dcoutts** [comment]
We can do this by decoding the Tokens to a Term and pretty printing the Term in CBOR syntax.

The testsuite already has a function `diagnosticNotation :: Term -> String` that produces the CBOR syntax (though it's for a type from the reference impl so some changes would be needed).

So yes, this should be straightforward.

**Comment by arianvp** [comment]
Something like: http://cbor.me/ ?

Which prints out these kind of useful messages:

```
a1       # map(1)
   61    # text(1)
      61 # "a"
   83    # array(3)
      01 # unsigned(1)
      03 # unsigned(3)
      04 # unsigned(4)

```

**Comment by thoughtpolice** [comment]
Why not both! We could have a `.Pretty` module with multiple functions! :)

**Comment by dcoutts** [comment]
I meant the latter form, which is informally specified in the CBOR spec, and which we already generate in the tests to compare to the reference test vectors.

**Comment by axman6** [comment]
I started implementing this, but got stuck when trying to use FlatTerm instead of Encoding directly. I'll give it another go this week if I find time. Current looks like this:

```haskell
pprintFlatTerm :: FlatTerm -> String
pprintFlatTerm tts = ppr 0 tts "" where
  ppr :: Int -> FlatTerm -> ShowS
  ppr _ [] = id
  ppr n (t:ts) = showString ("\n" ++ replicate n ' ') . case t of
    (TkInt x)      -> showString (if x < 0 then "# unsigned(" else "# signed(")
                      . showInt x
                      . showString ")"
                      . ppr n ts
    (TkInteger x)  -> showString "# "
                      . showString (if x < 0 then "unsigned(" else "signed(")
                      . showInt x
                      . showString ") (Integer)"
                      . ppr n ts
    (TkBytes x)    -> showString "# bytes("
                      . showInt (BS.length x)
                      . showString ")\n"
                      . showString (replicate (n+3) ' ' ++ "0x")
                      . hexBS x
                      . ppr n ts
    TkBytesBegin   -> showString "# array(*)"
                      . ppr (n+3) ts
    (TkString x)   -> showString "# text("
                      . showInt (T.length x)
                      . showString ")\n"
                      . showString (replicate (n+3) ' ')
                      . showString (show x)
                      . ppr n ts
    TkStringBegin  -> showString "# text(*)"
                      . ppr (n+3) ts
    (TkListLen x)  -> showString "# list("
                      . showInt x
                      . showString ")"
                      . ppr (n+3) ts
    TkListBegin    -> showString "# list(*)"
                      . ppr (n+3) ts
    (TkMapLen x)   -> showString "# map("
                      . showInt x
                      . showString ")"
                      . ppr (n+3) ts
    TkMapBegin     -> showString "# map(*)"
                      . ppr (n+3) ts
    TkBreak        -> showString "# primitive(*)"
                      . ppr (max 0 (n-3)) ts
    (TkTag x)      -> showString "# tag("
                      . showInt x
                      . showString ")"
                      . ppr (n+3) ts
    (TkBool x)     -> showString "# bool("
                      . shows x
                      . showString ")"
                      . ppr n ts
    TkNull         -> showString "# null" . ppr n ts
    (TkSimple x)   -> showString "# simple("
                      . showInt x
                      . showString ")"
                      . ppr n ts
    (TkFloat16 x)  -> showString "# float16("
                      . showFloat x
                      . showString ")"
                      . ppr n ts
    (TkFloat32 x)  -> showString "# float32("
                      . showFloat x
                      . showString ")"
                      . ppr n ts
    (TkFloat64 x)  -> showString "# float64("
                      . showFloat x
                      . showString ")"
                      . ppr n ts


hexBS :: ByteString -> ShowS
hexBS = foldr (.) id . map (\n -> (if n < 16 then ('0':) else id) . showHex n) . BS.unpack
```

**Comment by axman6** [comment]
Obviously this is missing the hex representation, which should be easier to produce with Encoding.

**Comment by thoughtpolice** [comment]
Closed by c1cafbeadb0a3ae11ce89580dc4e55f6677ead32. We'll do future work in other tickets if necessary.
